### ISSUE TITLE
Sub-task: 4.1 Unified Inference Engine Implementation

### ISSUE THREAD

**Issue by dhar174** [dhar174/steel_defect_demo#76]
### Overview
This sub-task focuses on the implementation of the `DefectPredictionEngine` class, which will serve as the core of the real-time inference system. This engine must be capable of loading, managing, and executing both the baseline (XGBoost) and deep learning (LSTM) models. It will unify the feature engineering, sequence processing, and prediction logic into a single, cohesive unit.

This is a sub-issue of #7.

### File Location
- `src/inference/inference_engine.py`

### Class Structure (`DefectPredictionEngine`)
```python
import numpy as np
import pandas as pd
from typing import Dict, List, Optional
import joblib
import torch
import yaml
from src.features.feature_engineer import CastingFeatureEngineer
from src.features.feature_extractor import SequenceFeatureExtractor
from src.models.baseline_model import BaselineXGBoostModel
from src.models.lstm_model import SteelDefectLSTM

class DefectPredictionEngine:
    """Unified inference engine for both models"""
    
    def __init__(self, config_path: str):
        # Load inference configuration
        pass
        
    def load_models(self, baseline_model_path: str, lstm_model_path: str) -> None:
        """Load trained models, feature engineer, and sequence processor."""
        pass
    
    def predict_baseline(self, time_series: pd.DataFrame) -> float:
        """Generate features and get prediction from baseline model."""
        pass
    
    def predict_lstm(self, time_series: pd.DataFrame) -> float:
        """Prepare sequence and get prediction from LSTM model."""
        pass
    
    def predict_ensemble(self, time_series: pd.DataFrame) -> Dict:
        """Get ensemble prediction with confidence and latency."""
        pass
```

#### 1. **Initialization (`__init__`)**
- Load and parse the `inference_config.yaml` file.
- Initialize attributes for models (`baseline_model`, `lstm_model`), the feature engineer (`feature_engineer`), and the sequence processor (`sequence_processor`) to `None`.
- Store relevant configuration details, such as model paths and ensemble weights.

#### 2. **Model Loading (`load_models`)**
- Implement logic to load the serialized baseline XGBoost model using `joblib`.
- Implement logic to load the saved PyTorch LSTM model state dictionary.
- Instantiate `CastingFeatureEngineer` and `SequenceFeatureExtractor` to be used for preprocessing incoming data.
- Handle potential errors during file loading (e.g., file not found, corrupted files).

#### 3. **Baseline Prediction (`predict_baseline`)**
- Integrate the `CastingFeatureEngineer` to transform the raw time-series data into a feature vector suitable for the baseline model.
- Use the loaded XGBoost model to make a prediction on the generated features.
- Return the resulting defect probability.

#### 4. **LSTM Prediction (`predict_lstm`)**
- Integrate the `SequenceFeatureExtractor` to normalize and pad/truncate the raw time-series data into a fixed-length sequence.
- Ensure the sequence is converted to a PyTorch tensor.
- Perform inference using the loaded LSTM model.
- Return the resulting defect probability.

#### 5. **Ensemble Prediction (`predict_ensemble`)**
- Orchestrate calls to both `predict_baseline` and `predict_lstm`.
- Implement a weighted average ensembling strategy based on weights defined in `inference_config.yaml`.
- Calculate a final, unified prediction probability.

#### 6. **Confidence and Uncertainty**
- Implement a method to calculate prediction confidence. This can be a simple function of the prediction's distance from the 0.5 threshold or a more advanced method if available (e.g., using dropout in the LSTM at inference time).

#### 7. **Performance Monitoring**
- Add internal logic to track the execution time (latency) for each prediction step (feature engineering, sequence processing, model inference).
- The `predict_ensemble` method should return a dictionary containing not only the prediction but also the latency and confidence score.

### Acceptance Criteria
- [ ] `DefectPredictionEngine` class is created in `src/inference/inference_engine.py`.
- [ ] `load_models` method successfully loads both the XGBoost and LSTM models from disk.
- [ ] `predict_baseline` correctly preprocesses data and returns a valid probability.
- [ ] `predict_lstm` correctly preprocesses data and returns a valid probability.
- [ ] `predict_ensemble` returns a dictionary containing the final prediction, confidence score, and latency.
- [ ] Ensemble weights from the configuration are correctly applied.
- [ ] Unit tests are created in `tests/test_inference.py` to validate each method in `DefectPredictionEngine`.
- [ ] The implementation is fully type-hinted and passes linting checks.

### Dependencies
```
joblib>=1.2.0
torch>=1.13.0
numpy>=1.23.0
pandas>=1.5.0
pyyaml>=6.0
```
